src/libraries/System.Private.Xml.Linq/tests/misc/LoadSaveAsyncTests.cs

```diff
@@ -202,5 +202,68 @@ public static IEnumerable<object[]> RoundtripOptions_MemberData
             }
         }
 
+        public static IEnumerable<object[]> IsAsync_SaveOptions_Data
+        {
+            get
+            {
+                foreach (bool isAsync in new[] { true, false })
+                    foreach (SaveOptions saveOptions in Enum.GetValues(typeof(SaveOptions)))
+                        yield return new object[] { isAsync, saveOptions };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(IsAsync_SaveOptions_Data))]
+        public async Task SaveAsync_CallsAsyncOnly_SaveSync_CallsSyncOnly(bool isAsync, SaveOptions saveOptions)
+        {
+            XDocument document = XDocument.Parse("<root>Test document async save</root>");
+            var element = new XElement("Test");
+            using (ForceSyncAsyncStream stream = new ForceSyncAsyncStream(isAsync))
+            {
+                if (isAsync)
+                {
+                    await document.SaveAsync(stream, saveOptions, CancellationToken.None);
+                    await element.SaveAsync(stream, saveOptions, CancellationToken.None);
+                }
+                else
+                {
+                    document.Save(stream);
+                    element.Save(stream);
+                }
+            }
+        }
+    }
+
+    public class ForceSyncAsyncStream : MemoryStream
+    {
+        private bool _isAsync;
+
+        public ForceSyncAsyncStream(bool async)
+        {
+            _isAsync = async;
+        }
+        public override void Flush()
+        {
+            Assert.False(_isAsync, "Sync operation not allowed");
```

Comment:
```suggestion
            Assert.False(_isAsync, "Async operation not allowed");
```

Reply:
Wanted to illustrate if this Sync `Flush` method is called for `Async` only stream report  "Sync operation not allowed" (similar condition checked in and same exception message thrown for async only streams in aspnetcore )

Reply:
What I mean that you're currently passing the assert only for sync streams (`_isAsync == false`) but the message says that Sync is not allowed

Reply:
Understand your point, but the assertion fail only when there were Async stream supposed to make async calls but calling this sync method, then we cannot say `Async not allowed`, how about changing `Sync method call not allowed for Async stream`?

Reply:
The message in XUnit asserts should report what condition it wanted, not what condition it saw, or what the recovery action is. It's backwards from exception messages.

So this could be

```C#
throw new InvalidOperationException("Sync operations are not allowed.");
```

or

```C#
Assert.False(_isAsync, "Stream is not in asynchronous mode when synchronous Flush is called");
```

The former will fail with

```
  The.Namespace.Qualified.Name.Of.This.Method [FAIL]
      InvalidOperationException: Sync operations are not allowed.
   at ...
```

The latter with

```
  Stream is not in asynchronous mode when synchronous Flush is called
    Expected: False
        Actual: True
```

The assert could also be double-negated:

```C#
Assert.True(!_isAsync, "Stream is in synchronous mode when synchronous Flush is called");
```

To produce

```
  Stream is in synchronous mode when synchronous Flush is called
    Expected: True
        Actual: False
```

Reply:
Makes sense, thanks